packages/ckeditor5-watchdog/src/editorwatchdog.ts

```diff
@@ -323,6 +361,89 @@ export default class EditorWatchdog<TEditor extends Editor = Editor> extends Wat
 	}
 }
 
+class EditorWatchdogInitPlugin extends Plugin {
+	private _data: EditorData;
+
+	constructor( editor: Editor ) {
+		super( editor );
+
+		this._data = editor.config.get( '_watchdogInitialData' )!;
+
+		editor.config.set( '_watchdogInitialData', null );
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public init(): void {
+		this.editor.data.on( 'init', evt => {
+			evt.stop();
+
+			this.editor.model.enqueueChange( writer => {
+				this._restoreEditorData( writer );
+			} );
+		}, { priority: priorities.highest - 1 } );
+	}
+
+	/**
+	 * Checks if the editor instance is a multi-root editor.
+	 */
+	private _isMultiRootEditor( editor: Editor ): editor is MultiRootEditor {
+		return ( editor as MultiRootEditor ).getRootsAttributes !== undefined;
+	}
+
+	/**
+	 * Restores the editor by setting all operations, attributes and markers.
+	 */
+	private _restoreEditorData( writer: Writer ): void {
+		const editor = this.editor!;
+
+		Object.entries( this._data! ).forEach( ( [ root, { content, markers, attributes } ] ) => {
+			const parsedNodes: Array<Node | Element> = JSON.parse( content );
+			const children = [];
+
+			if ( this._isMultiRootEditor( editor ) ) {
+				editor.addRoot( root );
+			}
+
+			const rootElement = editor.model.document.getRoot( root )!;
+
+			for ( const [ key, value ] of attributes ) {
+				writer.setAttribute( key, value, rootElement );
+			}
+
+			for ( const child of parsedNodes ) {
+				if ( 'name' in child ) {
+					// If child has name property, it is an Element.
+					children.push( Element.fromJSON( child ) );
+				} else {
+					// Otherwise, it is a Text node.
+					children.push( Text.fromJSON( child ) );
+				}
+			}
+
+			Object.entries( markers ).forEach( ( [ markerName, markerOptions ] ) => {
+				writer.addMarker( markerName, markerOptions );
+			} );
```

Comment:
Add content first, then markers. For a while, these markers are created on non-existing model tree. If there's a piece of code that would check something with markers right away, it will crash.